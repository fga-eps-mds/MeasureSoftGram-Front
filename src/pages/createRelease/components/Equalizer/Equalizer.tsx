import React from 'react';
import EqualizerSingleSlider from '@pages/createRelease/components/EqualizerSingleSlider';
import { Box } from '@mui/material';
import { useCreateReleaseContext } from '@pages/createRelease/context/useCreateRelease';
import undelineRemover from '@utils/undelineRemover';
import useEqualizer from './hook/useEqualizer';

import * as Styles from './styles';

interface EquilizerProps {
  selectedCharacteristics: string[];
}

export default function Equalizer({ selectedCharacteristics }: EquilizerProps) {
  const { characteristics, equalize, addDeltaToChanges, changes } = useEqualizer(selectedCharacteristics);

  const { preConfigCharacteristics, handleChangeForm } = useCreateReleaseContext();

  const handleChangeCommitted = (characteristicName: string, newValue: number) => {
    addDeltaToChanges(characteristicName, newValue);
    handleChangeForm('changes', changes);
  };

  return (
    <Box justifyContent="space-around" display="flex">
      <Box gap="24px" display="flex">
        {preConfigCharacteristics?.map((item, index) => {
          const char = characteristics.find((i) => i.key === item);

          let value;

          if (char) {
            value = char.value;
          }

          return (
            <EqualizerSingleSlider
              key={item}
              index={index + 1}
              name={undelineRemover(item)}
              value={value || 50}
              disabled={!selectedCharacteristics.includes(item)}
              onChange={(newValue) => equalize(item, newValue)}
              onChangeCommitted={(newValue) => handleChangeCommitted(item, newValue)}
            />
          );
        })}
      </Box>

      <Styles.Labels>
        {preConfigCharacteristics?.map((item, index) => (
          <p key={item} data-testid="label">
            C{index + 1}: {undelineRemover(item)}
          </p>
        ))}
      </Styles.Labels>
    </Box>
  );
}
