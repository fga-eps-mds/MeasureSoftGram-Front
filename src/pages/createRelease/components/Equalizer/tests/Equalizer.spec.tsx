import '@testing-library/jest-dom';

import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import Equalizer from '../Equalizer';

jest.mock('@pages/createRelease/context/useCreateRelease', () => ({
  useCreateReleaseContext: () => ({
    preConfigCharacteristics: ['reliability', 'maintainability'],
    handleChangeForm: () => {}
  })
}));

jest.mock('@mui/material', () => ({
  ...jest.requireActual('@mui/material'),
  Slider: ({ ...props }: any) => {
    const { id, name, min, max, onChangeCommitted, onChange } = props;
    return (
      <input
        data-testid={props['data-testid']}
        type="range"
        id={id}
        name={name}
        min={min}
        max={max}
        onChange={(event) => onChangeCommitted('_', event.target.value)}
        onInput={(event) => onChange('_', (event.target as HTMLInputElement).value)}
      />
    );
  }
}));

describe('<Equalizer />', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const renderEqualizer = () =>
    render(<Equalizer selectedCharacteristics={['usability', 'compatibility', 'security']} />);

  describe('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = renderEqualizer();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('Comportamento', () => {
    it('Deve chamar onChange ao movimnetar o slider', () => {
      const { getAllByTestId } = renderEqualizer();

      const sliders = getAllByTestId('single-slider');
      const labels = getAllByTestId('label');

      fireEvent.change(sliders[0], { target: { value: 10 } });
      fireEvent.input(sliders[0], { target: { value: 70 } });

      expect(sliders.length).toBe(2);
      expect(labels.length).toBe(2);
    });
  });
});
