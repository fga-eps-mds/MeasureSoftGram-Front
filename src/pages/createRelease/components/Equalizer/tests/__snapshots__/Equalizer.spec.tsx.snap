// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Equalizer /> Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-ckc6o3"
      >
        <div
          class="MuiBox-root css-t0lfuy"
        >
          <div
            class="MuiBox-root css-xi606m"
          >
            <div
              class="MuiBox-root css-vycneo"
            >
              <input
                data-testid="single-slider"
                max="100"
                min="0"
                type="range"
              />
            </div>
            <p
              class=""
              data-mui-internal-clone-element="true"
            >
              C
              1
            </p>
          </div>
          <div
            class="MuiBox-root css-xi606m"
          >
            <div
              class="MuiBox-root css-vycneo"
            >
              <input
                data-testid="single-slider"
                max="100"
                min="0"
                type="range"
              />
            </div>
            <p
              class=""
              data-mui-internal-clone-element="true"
            >
              C
              2
            </p>
          </div>
        </div>
        <div
          class="styles__Labels-sc-1fexhzx-0 kyDTDD"
        >
          <p
            data-testid="label"
          >
            C
            1
            : 
            reliability
          </p>
          <p
            data-testid="label"
          >
            C
            2
            : 
            maintainability
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-ckc6o3"
    >
      <div
        class="MuiBox-root css-t0lfuy"
      >
        <div
          class="MuiBox-root css-xi606m"
        >
          <div
            class="MuiBox-root css-vycneo"
          >
            <input
              data-testid="single-slider"
              max="100"
              min="0"
              type="range"
            />
          </div>
          <p
            class=""
            data-mui-internal-clone-element="true"
          >
            C
            1
          </p>
        </div>
        <div
          class="MuiBox-root css-xi606m"
        >
          <div
            class="MuiBox-root css-vycneo"
          >
            <input
              data-testid="single-slider"
              max="100"
              min="0"
              type="range"
            />
          </div>
          <p
            class=""
            data-mui-internal-clone-element="true"
          >
            C
            2
          </p>
        </div>
      </div>
      <div
        class="styles__Labels-sc-1fexhzx-0 kyDTDD"
      >
        <p
          data-testid="label"
        >
          C
          1
          : 
          reliability
        </p>
        <p
          data-testid="label"
        >
          C
          2
          : 
          maintainability
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
